### PR TITLE
[MIG] 12.0 sale_rental

### PR DESCRIPTION
**Pull request by gurneyalex**
extracted the sale_rental module migration from #899 

Depends on:
* [x] #943

### CODE REVIEW

File: sale_rental/__manifest__.py
Comment (Swapnesh-SerpentCS) [inline sale_rental/__manifest__.py]: Can we move `security/*` to first?

File: sale_rental/models/product.py
Comment (Swapnesh-SerpentCS) [inline sale_rental/models/product.py:38]: Can we have separate method at least wich return allowed list of UoM?
(Or Is it planned for next release?)

File: sale_rental/models/sale_order.py
Comment (Swapnesh-SerpentCS) [inline sale_rental/models/sale_order.py]:
```suggestion
    rental = fields.Boolean(string='Rental', default=False)
```

File: sale_rental/models/sale_order.py
Comment (Swapnesh-SerpentCS) [inline sale_rental/models/sale_order.py:100]:
```suggestion
            elif line.sell_rental_id and line.product_uom_qty != line.sell_rental_id.rental_qty:

```

File: sale_rental/models/sale_rental.py
Comment (Swapnesh-SerpentCS) [inline sale_rental/models/sale_rental.py]:
```suggestion
        'res.company', related='start_order_line_id.company_id',
```
See https://github.com/odoo/odoo/blob/12.0/addons/sale/models/sale.py#L1216

File: sale_rental/models/stock.py
Comment (Swapnesh-SerpentCS) [inline sale_rental/models/stock.py]:
```suggestion
    def _onchange_rental_allowed(self):
```

File: sale_rental/models/stock.py
Comment (Swapnesh-SerpentCS) [inline sale_rental/models/stock.py:41]: IMO we should not use` try:` here and use  `raise_if_not_found=False`

File: sale_rental/models/stock.py
Comment (Swapnesh-SerpentCS) [inline sale_rental/models/stock.py]: Just a comment, we are searching for record with `with_context(lang='en_US')` but creating new one with lang

**Comment by marylla** [comment]
Hi there, 
I tried to test this pull request in the runbot and forced its build. 
But I cannot install the module because the dependend module "sale_start_end_dates" is not available.
I checked the associated pull request #943 which is merged since 2 days. Yay. :)

Can someone check what has to be done here to make functional reviews with the runbot possible?
Thank you!

**Comment by pedrobaeza** [comment]
This pull request should be rebased

**Review by sadamo — APPROVED** [review]

**Comment by rafaelbn** [comment]
Hi @gurneyalex please could you rebase this PR? Dependency has been merged

**Comment by gurneyalex** [comment]
rebased

**Comment by gurneyalex** [comment]
But I thought the OCA bot would do the rebase before merging. Did I miss something?

File: sale_rental/__manifest__.py
Comment (gurneyalex) [reply sale_rental/__manifest__.py]: yes.

File: sale_rental/models/product.py
Comment (gurneyalex) [reply sale_rental/models/product.py:38]: I don't know. You need to discuss this with @alexis-via

File: sale_rental/models/sale_order.py
Comment (gurneyalex) [reply sale_rental/models/sale_order.py]: False is already the default for boolean fields.

File: sale_rental/models/sale_order.py
Comment (gurneyalex) [reply sale_rental/models/sale_order.py:100]: line length issue. I prefer keeping the original code as is.

File: sale_rental/models/stock.py
Comment (gurneyalex) [reply sale_rental/models/stock.py]: fixed

File: sale_rental/models/stock.py
Comment (sadamo) [inline sale_rental/models/stock.py]: The separation of push and pull rules are not needed since they are from the same object, just a "route_id" parameter in search is enough?

If the separation is needed, then in push_rule_to_delete, the domain fields does not exists in stock.rule object and should be refactored.

**Review by sadamo — CHANGES_REQUESTED** [review]

**Comment by sadamo** [comment]
@gurneyalex I found some issues in your migration and I made a PR to your branch to fix them.
https://github.com/gurneyalex/sale-workflow/pull/3

Specifically, if stock moves are created, then we can not delete the stock rules anymore. So, rather than delete, we archive the stock rules. 

Maybe it can help you. :wink:

**Review by marylla — APPROVED** [review]
I made a functional review according to the workflows shown in alexis' youtube video in the sale_rental module of odoo 7 (https://www.youtube.com/watch?v=9o0QrGryBn8).
I could configure the product and its rental service, the rental itself, rental extensions and the selling of the product. No errors or exceptions occured and the states of sale orders, rentals and delivery orders and the inventory locations were correct.
Thank you for porting the module! :)

**Review by yweng8111 — APPROVED** [review]
Thank you for your great work ! I made a Codereview. It looks very nice.

**Comment by marylla** [comment]
@alexis-via Could you please review this PR, so it can be merged soon? Thanks a lot.

**Comment by marylla** [comment]
@gurneyalex Can you please check the translation files? There might be some duplicate entries, which leads to problems when updating translations.
Thank you!

**Comment by EBII** [comment]
... need more time

**Comment by rousseldenis** [comment]
@gurneyalex Is this ready ?

**Comment by josepmy** [comment]
@gurneyalex @marylla is this ready or can I test and help?

**Comment by marylla** [comment]
The translation file still contains duplicated entries.

If I load a new language and overwrite existing terms, I get this exception:

<pre>
Odoo Server Error
Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/http.py", line 656, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/.repo_requirements/odoo/odoo/http.py", line 314, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/.repo_requirements/odoo/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/.repo_requirements/odoo/odoo/http.py", line 698, in dispatch
    result = self._call_function(**self.params)
  File "/.repo_requirements/odoo/odoo/http.py", line 346, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 519, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/OCB-12.0/addons/web/controllers/main.py", line 966, in call_button
    action = self._call_kw(model, method, args, {})
  File "/home/odoo/OCB-12.0/addons/web/controllers/main.py", line 954, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 759, in call_kw
    return _call_kw_multi(method, model, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 746, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/addons/base/wizard/base_language_install.py", line 36, in lang_install
    mods.with_context(overwrite=self.overwrite)._update_translations(self.lang)
  File "/.repo_requirements/odoo/odoo/addons/base/models/ir_module.py", line 885, in _update_translations
    self.env['ir.translation'].load_module_terms(mod_names, filter_lang)
  File "/.repo_requirements/odoo/odoo/addons/base/models/ir_translation.py", line 825, in load_module_terms
    tools.trans_load(self._cr, trans_file, lang, verbose=False, module_name=module_name, context=context)
  File "/.repo_requirements/odoo/odoo/tools/translate.py", line 1016, in trans_load
    result = trans_load_data(cr, fileobj, fileformat, lang, verbose=verbose, module_name=module_name, context=context)
  File "/.repo_requirements/odoo/odoo/tools/translate.py", line 1160, in trans_load_data
    irt_cursor.finish()
  File "/.repo_requirements/odoo/odoo/addons/base/models/ir_translation.py", line 156, in finish
    """ % (self._model_table, self._table))
  File "/.repo_requirements/odoo/odoo/sql_db.py", line 148, in wrapper
    return f(self, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/sql_db.py", line 225, in execute
    res = self._obj.execute(query, params)
psycopg2.ProgrammingError: ON CONFLICT DO UPDATE command cannot affect row a second time
HINT:  Ensure that no rows proposed for insertion within the same command have duplicate constrained values.
</pre>

It would be great if this can be fixed before the merge.

**Comment by gurneyalex** [comment]
@rousseldenis I just merged the PR from @sadamo, waiting for the CI

**Comment by rousseldenis** [comment]
@gurneyalex Could you squash ? Thanks!

**Comment by gurneyalex** [comment]
@rousseldenis rebased and squashed (in all the cases where this would not cause a conflict)

**Comment by gurneyalex** [comment]
@marylla I think that merging this will give weblate an opportunity to fix the sale_rental.pot file
@sbidoul do you confirm?

**Comment by rousseldenis** [comment]
> @marylla I think that merging this will give weblate an opportunity to fix the sale_rental.pot file
> @sbidoul do you confirm?

What was te problem with pot file ?

**Comment by gurneyalex** [comment]
> > @marylla I think that merging this will give weblate an opportunity to fix the sale_rental.pot file
> > @sbidoul do you confirm?
> 
> What was te problem with pot file ?

https://github.com/OCA/sale-workflow/pull/944#issuecomment-584047210

**Comment by sbidoul** [comment]
@gurneyalex When you merge, upon test success, travis will run `click-odoo-makepot` to generate and git push a clean `.pot` file. Then weblate should pick it up and update the `.po` files using `msgmerge -U`.

**Comment by josepmy** [comment]
@gurneyalex I cloned from https://github.com/gurneyalex/sale-workflow/tree/12.0-mig_sale_rental/sale_rental but seems the the rental tab don’t appear on products.

**Review by mliell — APPROVED** [review]

**Comment by josepmy** [comment]
Hi @pedrobaeza, The PR was merged?

**Comment by desdelinux** [comment]
@alexis-via Please, your review

**Review by HaraldPanten — APPROVED** [review]

**Review by sergiocorato — APPROVED** [review]
LGTM

**Review by marcelsavegnago — APPROVED** [review]

**Comment by sergiocorato** [comment]
@OCA/crm-sales-marketing-maintainers merge?

**Comment by marcelsavegnago** [comment]
> @gurneyalex When you merge, upon test success, travis will run `click-odoo-makepot` to generate and git push a clean `.pot` file. Then weblate should pick it up and update the `.po` files using `msgmerge -U`.

@sbidoul What could be the reason for this problem? The link did not work and I also did not find the component in the weblate.

https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_rental/
image

**Comment by sbidoul** [comment]
Ow, we have DEBUG=True in django settings ? Let me fix that first.

**Comment by sbidoul** [comment]
@marcelsavegnago there was a merge conflict in l10n-spain that blocked the weblate component generator. It's fixed now.

I'll now need to find out why nobody received an alert about that :/

**Comment by pedrobaeza** [comment]
So a problem in a repository blocks the rest of the updates?

**Comment by sbidoul** [comment]
It would seem so, yes. For the curious, it broke [here](https://github.com/acsone/weblate-odoo-component-generator/blob/a612a2ab861a91f6b30a2ac56bdf52b42c84cfd2/wocg/create_components.py#L90) because a manifest had a broken syntax following a git rebase error.

**Comment by pedrobaeza** [comment]
OK, thanks for the information. If this can be ironed for avoiding the total collapse, it would be great.

**Comment by marcelsavegnago** [comment]
Thank you guys.
image

**Pull request merged**
